### PR TITLE
Add `advanced_parquet_index.rs` example of index in into parquet files

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

Closes https://github.com/apache/datafusion/issues/10580

## Rationale for this change
See https://github.com/apache/datafusion/issues/10580

This shows how to use the APIs in  https://github.com/apache/datafusion/issues/9929

Building and using external indexes in DataFusion is an important feature. Adding an example of how to do so will help drive the design and APIs and help other people discover it more easiy

Specifically, this example illustrates how to configure the ParquetExec to read  only some row groups / pages *within* a file and shows how to avoid reading the metadata each time. This is required for low latency parquet query access

## What changes are included in this PR?

Add a new example: `parquet_index_advanced`

## Are there any user-facing changes?
This is an example

### CODE REVIEW

File: datafusion/common/src/column.rs [hunk 1]
```diff
@@ -127,6 +127,13 @@ impl Column {
             })
     }
 
+    /// return the column's name.
```
Comment (alamb) [inline datafusion/common/src/column.rs:130]: I added some small access APIs to make the example easier to read. I can revert these or put them in a different PR if reviewers prefer

**Review by hengfeiyang — APPROVED** [review]

File: datafusion-examples/examples/advanced_parquet_index.rs [hunk 2]
```diff
@@ -0,0 +1,602 @@
...
+    indexed_file: IndexedFile,
+    /// The underlying object store
+    object_store: Arc<dyn ObjectStore>,
+    /// if true, use row selections in addition to row group selections
+    use_row_selections: AtomicBool,
+}
+impl IndexTableProvider {
+    /// Create a new IndexTableProvider
+    /// * `dir` - the directory containing the parquet files
+    /// * `object_store` - the object store implementation to use for reading the files
```
Comment (adriangb) [inline datafusion-examples/examples/advanced_parquet_index.rs:177]: Seems like this is only taking an object store now and creating the directory internally

Comment (alamb) [reply datafusion-examples/examples/advanced_parquet_index.rs:177]: good call -- fixed in 79e1476a2

File: datafusion-examples/README.md [hunk 3]
```diff
@@ -45,6 +45,7 @@ cargo run --example csv_sql
 - [`advanced_udaf.rs`](examples/advanced_udaf.rs): Define and invoke a more complicated User Defined Aggregate Function (UDAF)
 - [`advanced_udf.rs`](examples/advanced_udf.rs): Define and invoke a more complicated User Defined Scalar Function (UDF)
 - [`advanced_udwf.rs`](examples/advanced_udwf.rs): Define and invoke a more complicated User Defined Window Function (UDWF)
+- ['advanced_parquet_index.rs'](examples/advanced_parquet_index.rs): Creates a detailed secondary index that covers the contents of several parquet files
```
Comment (Weijun-H) [inline datafusion-examples/README.md:48]:
```suggestion
- [`advanced_parquet_index.rs`](examples/advanced_parquet_index.rs): Creates a detailed secondary index that covers the contents of several parquet files
```

File: datafusion-examples/examples/advanced_parquet_index.rs [hunk 4]
```diff
@@ -0,0 +1,595 @@
...
+use std::collections::{HashMap, HashSet};
+use std::fs::File;
+use std::ops::Range;
+use std::path::{Path, PathBuf};
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+use tempfile::TempDir;
+use url::Url;
+
+/// This example demonstrates using low level DataFusion APIs to read ony
```
Comment (Weijun-H) [inline datafusion-examples/examples/advanced_parquet_index.rs:64]:
```suggestion
/// This example demonstrates using low level DataFusion APIs to read only
```

File: datafusion-examples/examples/advanced_parquet_index.rs [hunk 5]
```diff
@@ -0,0 +1,595 @@
...
+/// Depending on the information from the index, DataFusion can make a request
+/// to the storage service (e.g. S3)  to read only the necessary data.
+///
+/// Note that this example uses a hard coded index implementation. For a more
+/// realistic example of creating an index to prune files, see the
+/// `parquet_index.rs` example.
+///
+/// Specifically, this example illustrates how to:
+/// 1. Use [`ParquetFileReaderFactory`] to avoid re-reading parquet metadata on each query
+/// 2. Use [`PruningPredicate`] for predicate anaysis
```
Comment (Weijun-H) [inline datafusion-examples/examples/advanced_parquet_index.rs:84]:
```suggestion
/// 2. Use [`PruningPredicate`] for predicate analysis
```

File: datafusion-examples/examples/advanced_parquet_index.rs [hunk 6]
```diff
@@ -0,0 +1,595 @@
...
+            // evenly distributed between  in row groups, calculate in what row
+            // group this value appears and tell the parquet reader to read it
+            let val = *val as usize;
+            let num_rows_in_row_group = 1000 / plan.len();
+            let row_group_index = val / num_rows_in_row_group;
+            plan.scan(row_group_index);
+
+            // If we want to use row selections, which the parquet reader can
+            // use to skip data pages when the parquet file has a "page index"
+            // and the reader is configured to read it, add a row seelction
```
Comment (Weijun-H) [inline datafusion-examples/examples/advanced_parquet_index.rs:263]:
```suggestion
            // and the reader is configured to read it, add a row selection
```

File: datafusion-examples/examples/advanced_parquet_index.rs [hunk 7]
```diff
@@ -0,0 +1,595 @@
...
+///
+/// [`ListingTable`]: datafusion::datasource::listing::ListingTable
+#[tokio::main]
+async fn main() -> Result<()> {
+    // the object store is used to read the parquet files (in this case, it is
+    // a local file system, but in a real system it could be S3, GCS, etc)
+    let object_store: Arc<dyn ObjectStore> =
+        Arc::new(object_store::local::LocalFileSystem::new());
+
+    // Create a a custom table provider with our special index.
```
Comment (Weijun-H) [inline datafusion-examples/examples/advanced_parquet_index.rs:102]:
```suggestion
    // Create a custom table provider with our special index.
```

**Review by Weijun-H — APPROVED** [review]
Nice example! Thanks @alamb 👍

File: datafusion-examples/examples/advanced_parquet_index.rs [hunk 8]
```diff
@@ -0,0 +1,656 @@
...
+/// 3. Pass a row group selection to [`ParuetExec`]
+/// 4. Pass a row selection (within a row group) to [`ParuetExec`]
+///
+/// Note this is a *VERY* low level example for people who want to build their
+/// own custom indexes (e.g. for low latency queries). Most users should use
+/// higher level APIs for reading parquet files:
+/// [`SessionContext::read_parquet`] or [`ListingTable`], which also do file
+/// pruning based on parquet statistics (using the same underlying APIs)
+///
+/// # Diagram
```
Comment (alamb) [inline datafusion-examples/examples/advanced_parquet_index.rs:94]: I added a diagram here to try and explain what is going on visually for those so inclined

Comment (adriangb) [reply datafusion-examples/examples/advanced_parquet_index.rs:94]: Just curious, do you generate these by hand? I was making them by hand with auto complete assistance, but wondering if there's a better tool.

Comment (alamb) [reply datafusion-examples/examples/advanced_parquet_index.rs:94]: I use this https://monodraw.helftone.com/

 I think there are other online ascii art diagrammers, but that monodraw works well for me

**Comment by alamb** [comment]
This PR is just waiting for a committer to approve it so I can merge it in. @crepererum  or @tustvold  might you have a moment to do so?

**Comment by alamb** [comment]
@crepererum  do you by any chance have a few minutes to review and approve this PR?

**Comment by alamb** [comment]
I can't merge this PR until a committer approves it:

Screenshot 2024-06-21 at 12 47 52 PM

@Dandandan or @thinkharderdev  any chance you have some time to review an example (mostly comments)?

File: datafusion-examples/examples/advanced_parquet_index.rs [hunk 9]
```diff
@@ -0,0 +1,662 @@
...
+///          Parquet File                  decode the thrift footer
+///
+/// ```
+///
+/// Within a Row Group, Column Chunks store data in DataPages. This example also
+/// shows how to configure the ParquetExec to read a `RowSelection` (row ranges)
+/// which will skip unneeded data pages:
+///
+/// ```text
+///         ┌───────────────────────┐   If the RowSelection does not include any
```
Comment (thinkharderdev) [inline datafusion-examples/examples/advanced_parquet_index.rs:131]: Maybe add a note here that this can only happen if the parquet file has a `PageIndex`?

**Review by thinkharderdev — APPROVED** [review]
very nice!

File: datafusion-examples/examples/advanced_parquet_index.rs (hunk 9 above)
Comment (alamb) [reply datafusion-examples/examples/advanced_parquet_index.rs:132]: Good call -- done in d2f64777a

**Comment by alamb** [comment]
Thank you everyone for the great comments and feedback

**Pull request merged**